AEP/aep-00007.md

```diff
@@ -0,0 +1,225 @@
+{
+  /// <summary>
+  ///   Gets messages from the queue
+  /// </summary>
+  /// <param name="nbMessages">Number of messages to retrieve</param>
+  /// <param name="cancellationToken">Token used to cancel the execution of the method</param>
+  /// <returns>
+  ///   Enumerator allowing async iteration over the message queue
+  /// </returns>
+  IAsyncEnumerable<IQueueMessageHandler> PullMessagesAsync(int               nbMessages,
+                                                           CancellationToken cancellationToken = default);
+}
+```
+
+The `PullMessagesAsync` method returns an `IAsyncEnumerable<IQueueMessageHandler>` where the `IQueueMessageHandler` is the following interface.
+It represents the lifecycle of a message.
+The `QueueMessageStatus` is set by ArmoniK during the execution of the task.
+Then, the `DisposeAsync` method inherited from `IAsyncDisposable` is used to process the message from the queue by acknowledging it, not acknowledging it, or requeuing it depending on the status.
+
+```csharp
+/// <summary>
+///   Interface to handle queue messages lifecycle.
+/// </summary>
+public interface IQueueMessageHandler : IAsyncDisposable
+{
+  /// <summary>
+  ///   Used to signal that the message ownership has been lost
+  /// </summary>
+  [Obsolete("ArmoniK now manages loss of link with the queue")]
+  CancellationToken CancellationToken { get; set; }
+
+  /// <summary>
+  ///   Id of the message
+  /// </summary>
+  string MessageId { get; }
+
+  /// <summary>
+  ///   Task Id contained in the message
+  /// </summary>
+  string TaskId { get; }
+
+  /// <summary>
+  ///   Status of the message. Used when the handler is disposed to notify the queue.
+  /// </summary>
+  QueueMessageStatus Status { get; set; }
+
+  /// <summary>
+  ///   Date of reception of the message
+  /// </summary>
+  DateTime ReceptionDateTime { get; init; }
+}
+```
+
+```csharp
+/// <summary>
+///   Represents the status of a queue message
+/// </summary>
+public enum QueueMessageStatus
+{
+  /// <summary>
+  ///   Message is waiting for being processed.
+  /// </summary>
+  Waiting,
+
+  /// <summary>
+  ///   Message processing has failed. The message should be put back at the begin of the queue.
+  /// </summary>
+  Failed,
+
+  /// <summary>
+  ///   The message is being processed.
+  /// </summary>
+  Running,
+
+  /// <summary>
+  ///   Task is not ready to be processed. The message should be put at the end of the queue.
+  /// </summary>
+  Postponed,
+
+  /// <summary>
+  ///   The message has been processed. It can safely be removed from the queue.
+  /// </summary>
+  Processed,
+
+  /// <summary>
+  ///   The message processing has been cancelled. the message can safely be removed from the queue.
+  /// </summary>
+  Cancelled,
+
+  /// <summary>
+  ///   Message has been retried too many times and is considered as poisonous for the queue
+  /// </summary>
+  Poisonous,
+}
+```
+
+Push interface is a lot simpler.
```

Comment:
When I am reading it , I am a little bit confused since the previous part where saying that we do not a push in armonik. Perhaps it is clear for people knowing the in-out of armonik but it is confusing for someone else I think. Perhaps explicitely saying which process of the orchestrator is using it will help the comprehension (subtasking and resubmitt)

Reply:
There are 2 kinds of push:
- push messages from the queue to ArmoniK -> debate is up for this one.
- push messages from ArmoniK to the queue -> we cannot avoid this one.